#!/usr/bin/env python
PACKAGE = "plane_slam"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

size_type_enum = gen.enum([
                        gen.const("VGA",       int_t, 0, ""),
                        gen.const("QVGA",  int_t, 1, ""),
                        gen.const("QQVGA",     int_t, 2, "")
                       ], "Size of organized cloud as input.")

segment_enum = gen.enum([ gen.const("LineBased", int_t, 0, ""),
                        gen.const("Organized", int_t, 1, "")
                       ], "Method to segment plane.")


feature_type_enum = gen.enum([
                        gen.const("SURF", str_t, "SURF", ""),
                        gen.const("ORB", str_t, "ORB", ""),
                        gen.const("SIFT", str_t, "SIFT", ""),
                        gen.const("FAST", str_t, "FAST", "")
                       ], "Feature type.")

##
gen.add("cloud_size_type",  int_t, 0, "", 1, 0, 2, edit_method=size_type_enum)
gen.add("segment_method", int_t, 0, "Method to segment plane", 0, 0, 1, edit_method=segment_enum)
gen.add("feature_detector_type", str_t, 0, "", "ORB", edit_method = feature_type_enum)
gen.add("feature_extractor_type", str_t, 0, "", "ORB", edit_method = feature_type_enum)
gen.add("feature_good_match_threshold", double_t, 0, "", 4.0, 1.0, 10.0)
gen.add("feature_min_good_match_size", int_t, 0, "", 100, 0, 300)
gen.add("ransac_sample_size", int_t, 0, "", 4, 3, 10)
gen.add("ransac_iterations", int_t, 0, "", 200, 50, 500)
gen.add("ransac_min_inlier", int_t, 0, "", 50, 20, 200)
gen.add("ransac_inlier_max_mahal_distance", double_t, 0, "", 3.0, 0.1, 10.0)
gen.add("icp_max_distance", double_t, 0, "", 0.1, 0.02, 0.50)
gen.add("icp_iterations",    int_t,    0, "", 40,  5, 100)
gen.add("icp_tf_epsilon", double_t, 0, "", 1e-4, 1e-8, 1e-2)
gen.add("icp_min_indices",    int_t,    0, "Not used", 500,  100, 5000)
gen.add("icp_score_threshold",    double_t,    0, "", 0.3, 0.001, 0.8)
gen.add("pnp_iterations", int_t, 0, "", 200, 50, 500)
gen.add("pnp_min_inlier", int_t, 0, "", 50, 20, 200)
gen.add("pnp_repreject_error", double_t, 0, "", 1.2, 0.1, 10.0)
##
gen.add("display_input_cloud",  bool_t, 0,  "", False)
gen.add("display_line_cloud",  bool_t, 0,  "", False)
gen.add("display_normal",   bool_t, 0,  "", False)
gen.add("display_normal_arrow", bool_t, 0, "", False)
gen.add("display_plane",    bool_t, 0,  "", True)
gen.add("display_plane_number",    bool_t, 0,  "", True)
gen.add("display_plane_arrow",    bool_t, 0,  "", False)
gen.add("display_plane_inlier", bool_t, 0, "", True)
gen.add("display_plane_projected_inlier",    bool_t, 0,  "", False)
gen.add("display_plane_boundary", bool_t, 0, "", True)
gen.add("display_plane_hull",    bool_t, 0,  "", False)
gen.add("loop_one_message", bool_t, 0,  "", False)


exit(gen.generate(PACKAGE, "plane_slam", "PlaneSegment"))
